File: backend/routes/voters.js
const express = require('express');
const router = express.Router();
const pool = require('../config/db');
const { verifyAdmin } = require('../middleware/auth');

// Admin: list the roster. Shows who HAS voted (for turnout tracking)
// but there is no column or join anywhere that reveals WHAT they
// voted for — that link does not exist in the database.
router.get('/', verifyAdmin, async (req, res) => {
  try {
    const [rows] = await pool.query(
      'SELECT id, full_name, company_id, company_email, has_voted FROM voters ORDER BY full_name'
    );
    res.json(rows);
  } catch (err) {
    console.error(err);
    res.status(500).json({ message: 'Could not load voter roster' });
  }
});

// Admin: add an eligible voter to the roster
router.post('/', verifyAdmin, async (req, res) => {
  const { full_name, company_id, company_email } = req.body;
  if (!full_name || !company_id || !company_email) {
    return res.status(400).json({ message: 'Name, company ID, and company email are all required' });
  }

  try {
    const [existing] = await pool.query(
      'SELECT id FROM voters WHERE company_id = ? OR company_email = ?',
      [company_id.trim(), company_email.trim().toLowerCase()]
    );
    if (existing.length > 0) {
      return res.status(409).json({ message: 'That company ID or email is already on the roster' });
    }

    await pool.query(
      'INSERT INTO voters (full_name, company_id, company_email) VALUES (?, ?, ?)',
      [full_name.trim(), company_id.trim(), company_email.trim().toLowerCase()]
    );
    res.status(201).json({ message: 'Voter added to roster' });
  } catch (err) {
    console.error(err);
    res.status(500).json({ message: 'Could not add voter' });
  }
});

// Admin: remove someone from the roster (e.g. added by mistake)
router.delete('/:id', verifyAdmin, async (req, res) => {
  try {
    await pool.query('DELETE FROM voters WHERE id = ?', [req.params.id]);
    res.json({ message: 'Voter removed from roster' });
  } catch (err) {
    console.error(err);
    res.status(500).json({ message: 'Could not remove voter' });
  }
});

module.exports = router;

File: backend/app/routers/voters.py
from fastapi import APIRouter, Depends, HTTPException
from ..database import get_conn
from ..schemas import VoterCreate
from ..deps import get_current_admin

router = APIRouter(prefix="/api/voters", tags=["voters"])


@router.get("")
def list_voters(admin=Depends(get_current_admin)):
    """Shows who HAS voted (for turnout tracking) but there is no column or
    join anywhere that reveals WHAT they voted for — that link doesn't exist."""
    conn = get_conn()
    try:
        with conn.cursor() as cur:
            cur.execute(
                "SELECT id, full_name, company_id, company_email, has_voted FROM voters ORDER BY full_name"
            )
            return cur.fetchall()
    finally:
        conn.close()


@router.post("", status_code=201)
def add_voter(payload: VoterCreate, admin=Depends(get_current_admin)):
    company_id = payload.company_id.strip()
    company_email = str(payload.company_email).strip().lower()

    conn = get_conn()
    try:
        with conn.cursor() as cur:
            cur.execute(
                "SELECT id FROM voters WHERE company_id = %s OR company_email = %s",
                (company_id, company_email),
            )
            if cur.fetchone():
                raise HTTPException(status_code=409, detail="That company ID or email is already on the roster")

            cur.execute(
                "INSERT INTO voters (full_name, company_id, company_email) VALUES (%s,%s,%s)",
                (payload.full_name.strip(), company_id, company_email),
            )
        conn.commit()
        return {"message": "Voter added to roster"}
    finally:
        conn.close()


@router.delete("/{voter_id}")
def remove_voter(voter_id: int, admin=Depends(get_current_admin)):
    conn = get_conn()
    try:
        with conn.cursor() as cur:
            cur.execute("DELETE FROM voters WHERE id=%s", (voter_id,))
        conn.commit()
        return {"message": "Voter removed from roster"}
    finally:
        conn.close()
